Tmedia 468 identity block style button refactor pt1

JackHowa opened the pull request:
## Description
Reusable button refactor only for identity blocks

## Jira Ticket
- [TMEDIA-468](https://arcpublishing.atlassian.net/browse/TMEDIA-468)

## Acceptance Criteria

Match button style criteria: 
Primary
Primary reverse
Secondary
Secondary reverse
Default - for when a type isn't passed

- Match mocks for identity blocks and styles: 
![Screen Shot 2021-09-22 at 11 18 23](https://user-images.githubusercontent.com/5950956/134382513-f367bc42-e383-419f-ad6d-d4936e730c72.png)
![Screen Shot 2021-09-22 at 11 18 34](https://user-images.githubusercontent.com/5950956/134382538-ed02c02a-0331-4997-9ac4-e2b06c0fbe5b.png)
![Screen Shot 2021-09-22 at 11 18 57](https://user-images.githubusercontent.com/5950956/134382581-c89753fe-93ed-4b94-8979-158d60632590.png)

## Test Steps

1. Checkout this branch `git checkout TMEDIA-468-identity-block-style-button-refactor-pt1`
2. Run fusion repo with linked blocks `npx fusion start -f -l @wpmedia/identity-block,@wpmedia/shared-styles`
3. Open pages for sign in with various primary color and nav bar background sites 
4. Go to mobile version by dragging each site
5. Sign in for a user and see the account button 
6. Drag in and see the mobile version of each site

demo-homepage: 
http://localhost/pf/demo-homepage/?_website=the-sun -> primary-color background
http://localhost/pf/demo-homepage/?_website=dagen -> light nav
http://localhost/pf/demo-homepage/?_website=the-prophet -> dark nav

## Effect Of Changes
### Before
- Using wrong color for icons in some cases
- Different naming for icons 

### After
<img width="1116" alt="Screen Shot 2021-09-20 at 10 11 28" src="https://user-images.githubusercontent.com/5950956/134382743-fd0e9908-17e7-48c5-ab49-e1d81a0bc4eb.png">
<img width="691" alt="Dagen" src="https://user-images.githubusercontent.com/5950956/134382767-d220626c-3eee-4a12-92d2-ddd13b2e7491.png">

## Dependencies or Side Effects
- Working on pt 2 for header nav buttons such as querly and sections
- Updated the form input to have the other button name updated

## Author Checklist
_The author of the PR should fill out the following sections to ensure this PR is ready for review._
- [x] Confirmed all the test steps a reviewer will follow above are working.
- [x] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [x] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [x] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [x] Confirmed this PR has unit test files
  - [x] Ran `npm run test`, made sure all tests are passing
  - [x] If the amount of work to write unit tests for this change are excessive,
please explain why (so that we can fix it whenever it gets refactored).
- [x] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist
_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] Linting code actions have passed.
- [ ] Ran the code locally based on the test instructions.
  - [ ] I don’t think this is needed to be tested locally. For example, a padding style change (storybook?) or a logic change (write a test).
- [ ] I am a member of the engine theme team so that I can approve and merge this. If you're not on the team, you won't have access to approve and merge this pr.
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.

Commit:
ci: Add background and addon essentials for handling background colors other than white

Commit:
refactor: Rename button styles based on new designs and vocab

Commit:
feat: Add get theme style button handler and logic for testing

Commit:
refactor: Remove unused props

Commit:
style: Remove unused type

Commit:
fix: Use consistent white color hex

nschubach commented on blocks/shared-styles/_children/button/index.jsx:
```diff
@@ -29,9 +34,27 @@ export const BUTTON_TYPES = {
+  switch (buttonStyle) {
+    case BUTTON_STYLES.PRIMARY:
+    case BUTTON_STYLES.SECONDARY_REVERSE:
+      iconColor = '#fff';
```
We don't have variable values for this?

JackHowa replied:
for the white color? I can update to use a constant

JackHowa replied:
updated @nschubach

nschubach replied:
It's not about the constant.  It's about the hardcoded value here.  Would this not be the inverse of the background?  Like, if the background is `primary`, then the icon would be `primary_reverse`?

nschubach replied:
In this case, is this the icon supposed to be the Secondary (foreground) color?

JackHowa replied:
<img width="770" alt="Screen Shot 2021-09-22 at 14 43 32" src="https://user-images.githubusercontent.com/5950956/134411108-33aeb942-430c-422e-9a9d-2d9b3bd7de25.png">
I'm going off of this

nschubach commented on blocks/shared-styles/_children/button/utils/getButtonStyle.js:
```diff
@@ -0,0 +1,22 @@
+}
+
+function getNavSpecificSecondaryButtonTheme(navColor, navBarBackground) {
+  if (navBarBackground === 'primary-color' || navColor === 'dark') {
```
What happens if the primary is not 'dark', but lighter color?  This kind of goes hand in hand with the other comment I put here.  If we have a Primary-color and a primary-color-reverse I would imagine those would be inverse for the background:color relationship.

matthewroach replied:
To follow up from slack - Primary button in the current implement is assumed to use the primary color but that will not always be the case. Primary Button and Primary Color are different - the primary button is the design it just so happens the base implementation uses the primary color - longer term the color used in primary button could be any color decided by a variable

matthewroach replied:
We currently have the same issue with the result list and search result list block "see more" button - it's an issue with the blocks assuming primary color is going to work with the predefined font color

Commit:
refactor: Update color variable names based on feedback

https://github.com/WPMedia/fusion-news-theme-blocks/pull/1098#discussion_r714245196

nschubach reviewed (dismissed):
little worried about the use of hard color values and the distinguishing between background/foreground colors.  I know that thematic concepts are being worked out, but I worry about using light/dark in regards to primary/secondary.  In the past I've used tools to pick readable colors(TinyColor) based on inputs, but I know we are not there yet.

If these are not a concern by Matt, I will concede until we get the theme stuff sorted out.

Commit:
Merge branch 'canary' into TMEDIA-468-identity-block-style-button-refactor-pt1

matthewroach commented on blocks/shared-styles/_children/button/index.jsx:
```diff
@@ -29,9 +34,32 @@ export const BUTTON_TYPES = {
+  }
+
   if (iconTypeString) {
     switch (iconTypeString) {
```
This switch statement is going to get pretty long over time, and the case parts are near on the same the only different is the Icon name - we could use an object map for handling returning the correct icon based on name

JackHowa replied:
ok updated

JackHowa replied:
updated this based on question

matthewroach commented on blocks/shared-styles/_children/button/index.story.jsx:
```diff
@@ -9,77 +9,177 @@ export default {
+  <Button
+    buttonStyle={BUTTON_STYLES.PRIMARY}
+    buttonType={BUTTON_TYPES.LABEL_AND_ICON}
+    iconType="user"
```
We are mixing how we specify prop values, buttonStyle and buttonType are using constants, but iconType and secondaryIconType is strings - We should talk about this at the next tech grooming

JackHowa replied:
yeah good point. should I make any changes in this pr for this? I will add this example to tech grooming

matthewroach commented on blocks/shared-styles/_children/button/index.story.jsx:
```diff
@@ -9,77 +9,177 @@ export default {
+
+export const LongText = () => (
+  <Button
+    text="Sign up to get the very best experience and this other really long work and such and more and even more text.Sign up to get the very best experience and this other really long work and such and more and even more text."
```
While this is possible - should we advocate for it? This would be valid in an `a` link button style but not an action `button`

JackHowa replied:
I can remove it

JackHowa replied:
updated

matthewroach commented on blocks/shared-styles/index.js:
```diff
@@ -16,6 +16,10 @@ import SmallPromoPresentation from './_children/promos/small';
+import {
+  getNavSpecificSecondaryButtonTheme,
+  getNavSpecificPrimaryButtonTheme,
+} from './_children/button/utils/getButtonStyle';
```
As we start to think about foundations and enhanced styling this feels like we wouldn't want this here, but more along side the block implementation as it has a very specific use case

JackHowa replied:
yeah I thought about that too. I put it here because this logic will be shared between two blocks -- identity and header nav chain block

JackHowa replied:
I'm thinking a sort of arc-utils package would be best for arc-specific logic

matthewroach replied:
That makes sense - I forgot about the header nav stuff - I was just thinking in terms of this PR with regards to that - then I think it's fine to be here for now

Commit:
docs: Remove long text story per feedback

https://github.com/WPMedia/fusion-news-theme-blocks/pull/1098#discussion_r714771936

Commit:
Merge branch 'TMEDIA-468-identity-block-style-button-refactor-pt1' of github.com:WPMedia/fusion-news-theme-blocks into TMEDIA-468-identity-block-style-button-refactor-pt1

Commit:
refactor: Use object rather than case switch for rendering icons

Commit:
refactor: Update the icon type string and logic

matthewroach commented on blocks/shared-styles/_children/button/index.jsx:
```diff
@@ -260,7 +276,7 @@ Button.propTypes = {
 
```
The `iconType` prop types should be a `PropTypes.oneOf(['user', 'chevron-up', 'chevron-down'])`

JackHowa replied:
updated!

Commit:
fix: Fix proptypes for icontype

Commit:
fix: Update default prop for icon type to prevent warnings

matthewroach reviewed (changes requested):
Couple of small things

1. The new Subscriptions Dialog component needs updated to reflect changes in this PR
2. The story custom ariaLabel isn't outputting the `aria-label` - Do we need this story? The label is coded to only output if it's an icon only button

Commit:
docs: Remove custom aria label story

Commit:
fix: Update fix for button style

JackHowa commented:
Updated -- thanks @matthewroach

JackHowa commented:
Removed aria label custom story
